pkg/model/awsmodel/firewall.go

```diff
@@ -297,6 +301,129 @@ func (b *FirewallModelBuilder) buildMasterRules(c *fi.ModelBuilderContext, nodeG
 	return masterGroups, nil
 }
 
+func (b *FirewallModelBuilder) FindDeletions(context *fi.ModelBuilderContext, cloud fi.Cloud) error {
+	awsCloud := cloud.(awsup.AWSCloud)
+
+	sgs, err := b.getClusterSecurityGroups(awsCloud)
+	if err != nil {
+		return err
+	}
+	for _, sg := range sgs {
+		name := fi.StringValue(sg.GroupName)
+		sgTask := context.Tasks["SecurityGroup/"+name].(*awstasks.SecurityGroup)
+
+		klog.V(4).Infof("found group %q with id %s", name, fi.StringValue(sg.GroupId))
+		// We assume that if the security group name ends with the cluster name it is owned by kOps.
+		// We must ignore security groups about which we don't know anything, as we cannot make tasks of them.
+		if !strings.HasSuffix(name, "."+b.ClusterName()) {
+			klog.V(4).Infof("skipping EC2 security group %q", name)
+			continue
+		}
+		for _, rule := range sg.IpPermissions {
+			ts := b.createRulesFromPermissions(sgTask, rule, false, sgs)
+			for _, task := range ts {
+				if _, ok := context.Tasks["SecurityGroupRule/"+*task.Name]; !ok {
+					if !featureflag.DeleteUnknownSGRules.Enabled() {
+						klog.Warningf("found unknown security group permission: %v", fi.DebugAsJsonString(task))
+						klog.Warning("you can delete this rule by using the DeleteUnknownSGRules feature gate")
+						klog.Warning(("a future version of kOps will remove this automatically"))
+					} else {
+						context.AddTask(task)
+					}
+				}
+			}
+		}
+		for _, rule := range sg.IpPermissionsEgress {
+			ts := b.createRulesFromPermissions(sgTask, rule, true, sgs)
+			for _, task := range ts {
+				if _, ok := context.Tasks["SecurityGroupRule/"+*task.Name]; !ok {
+					if !featureflag.DeleteUnknownSGRules.Enabled() {
+						klog.Warningf("found unknown security group permission: %v", fi.DebugAsJsonString(task))
+						klog.Warning("you can delete this rule by using the DeleteUnknownSGRules feature gate")
+						klog.Warning(("a future version of kOps will remove this automatically"))
+					} else {
+						context.AddTask(task)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (b *FirewallModelBuilder) getClusterSecurityGroups(cloud awsup.AWSCloud) (map[string]*ec2.SecurityGroup, error) {
+	sgs := make(map[string]*ec2.SecurityGroup)
+
+	request := &ec2.DescribeSecurityGroupsInput{
+		Filters: []*ec2.Filter{
+			awsup.NewEC2Filter("tag:KubernetesCluster", b.ClusterName()),
```

Comment:
maybe we filter on the owned tag too? that might be specific enough to not need to check the SG name's suffix matching the cluster name above.

https://github.com/kubernetes/kops/blob/fc91d0d459d28630157fd49a370525b482153482/tests/integration/update_cluster/complex/kubernetes.tf#L803